### PR TITLE
Prefer KVO for app storage observation

### LINKED ISSUE #3311: Crash when using @Shared appStorage during certain conditions
### Description

I'm seeing a crash with a similar backtrace to https://github.com/pointfreeco/swift-composable-architecture/issues/3247, but appears not to be caused by a thread issue. 

I'm using a @Shared appStorage key roughly like:

```swift
@Shared(.appStorage("key")) var isEnabled: Bool = true
```

When launching the app it crashes. I'm still working on a simplified repro but having a hard time isolating the exact code causing the problem, but thought I'd post the issue in case anyone else has an idea.

Screenshot 2024-08-26 at 12 24 36 PM

If you look at the backtrace, it appears an `NSAttributedString` is performing layout (which isn't in my code) which triggers an `NSUserDefault` update which in turn causes the `@Shared` storage to update immediately during the same render cycle?The only relevant logs related to the crash are:

```
precondition failure: setting value during update: 219864
```

Screenshot 2024-08-26 at 12 25 55 PM

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

_No response_

### Actual behavior

_No response_

### Steps to reproduce

_No response_

### The Composable Architecture version information

1.13.1

### Destination operating system

iOS 17.5

### Xcode version information

Xcode 15.4

### Swift Compiler version information

_No response_

### PR DESCRIPTION
**Pull request by stephencelis**
Observing the firehose notification from Notification Center is not only less performant, but it can cause crashes in SwiftUI if a notification is posted during a view body update, which can occur in certain situations.

Instead, we will prefer KVO unless we detect an invalid key, such as one that contains a `.` or is prefixed with an `@`.

Fixes #3311.

### CODE REVIEW

File: Sources/ComposableArchitecture/SharedState/PersistenceKey/AppStorageKey.swift [hunk 1]
```diff
@@ -290,42 +290,77 @@ extension AppStorageKey: PersistenceKey {
...
+      store.wrappedValue.addObserver(observer, forKeyPath: key, options: .new, context: nil)
+      removeObserver = { store.wrappedValue.removeObserver(observer, forKeyPath: key) }
     }
     let willEnterForeground: (any NSObjectProtocol)?
     if let willEnterForegroundNotificationName {
       willEnterForeground = NotificationCenter.default.addObserver(
         forName: willEnterForegroundNotificationName,
         object: nil,
-        queue: nil
+        queue: .main
```
Comment (stephencelis) [inline Sources/ComposableArchitecture/SharedState/PersistenceKey/AppStorageKey.swift:334]: I _believe_ this forces a thread hop, which means we should avoid the crash even on invalid KVO keys, but we could also do an explicit dispatch if we're not confident here.

File: Sources/ComposableArchitecture/SharedState/PersistenceKey/AppStorageKeyPathKey.swift [hunk 2]
```diff
@@ -60,22 +60,6 @@ extension AppStorageKeyPathKey: PersistenceKey, Hashable {
       observer.invalidate()
     }
   }
-
-  private class Observer: NSObject {
```
Comment (stephencelis) [inline Sources/ComposableArchitecture/SharedState/PersistenceKey/AppStorageKeyPathKey.swift:64]: This observer wasn't used anymore.

File: Makefile [hunk 3]
```diff
@@ -36,11 +38,17 @@ endif
 
 TEST_RUNNER_CI = $(CI)
 
-xcodebuild:
+warm-simulator:
+	@test "$(PLATFORM_ID)" != "" \
+		&& xcrun simctl boot $(PLATFORM_ID) \
+		&& open -a Simulator --args -CurrentDeviceUDID $(PLATFORM_ID) \
+		|| exit 0
```
Comment (stephencelis) [inline Makefile:45]: Trying to see if we can reduce CI flakiness by warming the simulators.

**Comment by denysdanyliukboosters** [comment]
@stephencelis Looks like this also fixes [this issue](https://github.com/pointfreeco/swift-composable-architecture/issues/3462).

**Comment by ddanilyuk** [comment]
@stephencelis Looks like this also fixes [this issue](https://github.com/pointfreeco/swift-composable-architecture/issues/3462).

**Pull request merged**